yapapi/executor/services.py

```diff
@@ -0,0 +1,411 @@
+            # potential solution:
+            # * use `aiostream.stream.merge`
+
+            ctl = instance.get_control_signal()
+            if ctl == ControlSignal.stop:
+                if instance.state == ServiceState.running:
+                    instance.service_state.stop()
+                else:
+                    instance.service_state.terminate()
+
+                logger.debug(f"{instance.service} state changed to {instance.state.value}")
+
+                handler = self._get_handler(instance)
+                batch = None
+
+        logger.info(f"{instance.service} decomissioned")
+
+    async def spawn_instance(self):
+        spawned = False
+
+        async def start_worker(agreement: rest.market.Agreement, node_info: NodeInfo) -> None:
+            nonlocal spawned
+            self.emit(events.WorkerStarted(agr_id=agreement.id))
+            try:
+                act = await self._engine.create_activity(agreement.id)
+            except Exception:
+                self.emit(
+                    events.ActivityCreateFailed(
+                        agr_id=agreement.id, exc_info=sys.exc_info()  # type: ignore
+                    )
+                )
+                self.emit(events.WorkerFinished(agr_id=agreement.id))
+                raise
+
+            async with act:
+                spawned = True
+                self.emit(events.ActivityCreated(act_id=act.id, agr_id=agreement.id))
+                self._engine.approve_agreement_payments(agreement.id)
+                work_context = WorkContext(
+                    act.id, node_info, self._engine.storage_manager, emitter=self.emit
+                )
+                task_id = f"{self.id}:{next(self._task_ids)}"
+                self.emit(
+                    events.TaskStarted(
+                        agr_id=agreement.id,
+                        task_id=task_id,
+                        task_data=f"Service: {self._service_class.__name__}"
+                    )
+                )
+
+                try:
+                    instance_batches = self._run_instance(work_context)
+                    try:
+                        await self._engine.process_batches(agreement.id, act, instance_batches)
+                    except StopAsyncIteration:
+                        pass
+
+                    self.emit(
+                        events.TaskFinished(
+                            agr_id=agreement.id,
+                            task_id=task_id,
+                        )
+                    )
+                    self.emit(events.WorkerFinished(agr_id=agreement.id))
+                except Exception:
+                    self.emit(
+                        events.WorkerFinished(
+                            agr_id=agreement.id, exc_info=sys.exc_info()  # type: ignore
+                        )
+                    )
+                    raise
+                finally:
+                    await self._engine.accept_payment_for_agreement(agreement.id)
```

Comment:
I would add the following here:
```
await self._job.release_agreement(agreement.id, allow_reuse=False)
```
in order to ensure that the agreement will be terminated soon after `start_worker()` exits.

Otherwise, the agreement will not be terminated until the provider terminates it after 90s of inactivity. And only then an invoice will be issued by the provider.

Reply:
Fixed in https://github.com/golemfactory/yapapi/pull/383/commits/dbefecd9f372beb0f5e09df2a6d70b41ca5634d2

Reply:
thank you @azawlocki